lang/cs/Org.Apache.REEF.IO/PartitionedData/FileSystem/FilePartition.cs

```diff
@@ -0,0 +1,191 @@
+                {
+                    File.Delete(localFilePath);
+                    Logger.Log(Level.Warning, "localFile already exists, delete it: " + localFilePath);
+                }
+
+                _fileSystem.CopyToLocal(sourceUri, localFilePath);
+                if (File.Exists(localFilePath))
+                {
+                    Logger.Log(Level.Info, string.Format
+                        (CultureInfo.CurrentCulture, "File {0} is Copied to local {1}.", sourceUri, localFilePath));
+                }
+                else
+                {
+                    string msg = string.Format
+                        (CultureInfo.CurrentCulture, "File {0} is NOT Copied to local {1}.", sourceUri, localFilePath);
+                    Exceptions.Throw(new FileLoadException(), msg, Logger);
+                }
+            }
+        }
+
+        //private string CopyFromRemote(string sourceFilePath)
+        //{
+        //    Logger.Log(Level.Info, "GetUriPrefix: " + _fileSystem.UriPrefix);
+
+        //    Uri sourceUri = new Uri(_fileSystem.UriPrefix + sourceFilePath);
+        //    Logger.Log(Level.Info, "remoteUri: " + sourceUri);
+
+        //    if (!_fileSystem.Exists(sourceUri))
+        //    {
+        //        throw new FileNotFoundException(string.Format(CultureInfo.CurrentCulture, "Remote File {0} does not exists.", sourceUri));
+        //    }
+
+        //    var localFilePath = Path.GetTempPath() + "-partition-" + DateTime.Now.ToString("yyyyMMddHHmmssfff");
+
+        //    Logger.Log(Level.Info, "localFile: " + localFilePath);
+
+        //    if (File.Exists(localFilePath))
+        //    {
+        //        File.Delete(localFilePath);
+        //        Logger.Log(Level.Warning, "localFile already exists, delete it: " + localFilePath);
+        //    }
+
+        //    _fileSystem.CopyToLocal(sourceUri, localFilePath);
+        //    if (File.Exists(localFilePath))
+        //    {
+        //        Logger.Log(Level.Info, string.Format
+        //            (CultureInfo.CurrentCulture, "File {0} is Copied to local {1}.", sourceUri, localFilePath));
+        //    }
+        //    else
+        //    {
+        //        Logger.Log(Level.Info, string.Format
+        //            (CultureInfo.CurrentCulture, "File {0} is not Copied to local {1}.", sourceUri, localFilePath));
+        //    }
+
+        //    return localFilePath;
+        //}
+
+        ~FilePartition()
+        {
+            Dispose(false);
+        }
+
+        public void Dispose() { Dispose(true); }
+        private void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                GC.SuppressFinalize(this);
+            }
+
+            if (_localFileNames != null)
+            {
+                foreach (var fileName in _localFileNames)
+                {
+                    File.Delete(fileName);
+                }
+                _localFileNames = null;
+            }
+        }
```

Comment:
This is a bad dispose pattern. It is sad that .NET documentations have not been updated since .NET 2.0 when this was the recommended pattern.

The right way of doing this:
 * Don't write destructor
 * No private Dispose(bool)
 * Just one public void Dispose(void) which is preferably thread safe.